src/logic/map_objects/tribes/ship.cc

```diff
@@ -883,23 +909,32 @@ WalkingDir Ship::get_scouting_direction() const {
 /// @note only called via player command
 void Ship::exp_construct_port(Game& game, const Coords& c) {
 	assert(expedition_);
-	const MapObject* mo = game.map().get_fcoords(c).field->get_immovable();
-	if ((mo != nullptr) && mo->descr().type() >= MapObjectType::BUILDING) {
-		// Another expedition ship (or an enemy player) was a second faster
-		set_ship_state_and_notify(
-		   ShipStates::kExpeditionWaiting, NoteShip::Action::kDestinationChanged);
-		return;
+	// recheck ownership before setting the csite
+	Map* map = game.mutable_map();
+	if (!expedition_->seen_port_buildspaces.empty()) {
+		const FCoords fc = map->get_fcoords(expedition_->seen_port_buildspaces.front());
+		if (!can_build_port_here(get_owner()->player_number(), *map, fc)) {
+			set_ship_state_and_notify(
+			   ShipStates::kExpeditionWaiting, NoteShip::Action::kDestinationChanged);
+			send_message(
+			   game, _("Port Space"), _("Port Space Lost"),
+			   _("A discovered port build space is not available for building a port anymore."),
+			   "images/wui/editor/fsel_editor_set_port_space.png");
+			expedition_->seen_port_buildspaces.clear();
```

Comment:
Please make this a function to avoid duplication.

Reply:
this would need to be a function of the ship, otherwise a lot of values won't be available in the function and we would need to call the funtion with approx 6 parameters.  
although I favor avoiding duplication in general I simply deem this not enough for an own ship function and not worth creating a function with 6 params. So I think we leave it like it is

Reply:
I think this is long enough for a private function for `Ship`. Now you even have the log message in addition to the 2 new translatable strings. Plus 2 levels of `if`s and a local variable. The functionality is also quite important.